bindings/matrix-sdk-ffi/src/room_preview.rs

```diff
@@ -1,9 +1,87 @@
-use matrix_sdk::{room_preview::RoomPreview as SdkRoomPreview, RoomState};
+use matrix_sdk::{
+    room_preview::{
+        JoinRoomPreviewAction as SdkJoinRoomPreviewAction, LeaveRoomPreviewAction as SdkLeaveRoomPreviewAction, RoomPreview as SdkRoomPreview,
+        RoomPreviewActions,
+    },
+    Client, RoomState,
+};
 use ruma::space::SpaceRoomJoinRule;
 
-/// The preview of a room, be it invited/joined/left, or not.
+use crate::{client::JoinRule, error::ClientError, room::Membership};
+
 #[derive(uniffi::Record)]
 pub struct RoomPreview {
+    pub info: RoomPreviewInfo,
+    pub room_preview_actions: RoomPreviewActions,
+    pub(crate) sdk_info: matrix_sdk::room_preview::RoomPreview,
+}
+
+#[derive(uniffi::Enum)]
+pub enum RoomPreviewAction {
+    Invited { join: JoinRoomPreviewAction, leave: LeaveRoomPreviewAction },
+    Knocked { leave: LeaveRoomPreviewAction },
+}
+
+#[derive(uniffi::Object)]
+pub struct JoinRoomPreviewAction {
+    inner: SdkJoinRoomPreviewAction,
+}
+
+#[matrix_sdk_ffi_macros::export]
+impl JoinRoomPreviewAction {
+    async fn run(&self) {
+        self.inner.run().unwrap()
+    }
+}
+
+impl From<matrix_sdk::room_preview::RoomPreviewActions> for RoomPreviewAction {
+    fn from(actions: RoomPreviewActions) -> Self {
+        match actions {
+            RoomPreviewActions::Invited { join, leave } => Self::Invited { join, leave },
+            RoomPreviewActions::Knocked { leave } => Self::Knocked { leave },
+        }
+    }
+}
+
+struct
+
+impl RoomPreview {
+    pub(crate) fn try_from_sdk(info: SdkRoomPreview, client: Client) -> Result<Self, ClientError> {
+        let info = RoomPreviewInfo {
+            room_id: info.room_id.to_string(),
+            canonical_alias: info.canonical_alias.map(|alias| alias.to_string()),
+            name: info.name,
+            topic: info.topic,
+            avatar_url: info.avatar_url.map(|url| url.to_string()),
+            num_joined_members: info.num_joined_members,
+            room_type: info.room_type.map(|room_type| room_type.to_string()),
+            is_history_world_readable: info.is_world_readable,
+            membership: info.state.map(|state| state.into()).unwrap_or_else(|| Membership::Left),
```

Comment:
nit: `unwrap_or` is sufficient; `unwrap_or_else` is to be used when some computation needs to happen in the closure.

Why can we assume `Left` in the first place, though?

Reply:
That's what `RoomPreview` does at the SDK level [here](https://github.com/matrix-org/matrix-rust-sdk/blob/b6e9d4f6bdb35b4f5c5320c63133b0a33d278c21/crates/matrix-sdk/src/room_preview.rs#L114) for join rules, so I assumed we could do the same. I'll change it to `Option<Membership>` instead.

Reply:
Handled in https://github.com/matrix-org/matrix-rust-sdk/commit/7bfb1bd1c4d3bd65f15b132330f3dbc9189063f0.